### DISCUSSION TITLE
useFieldArray - remove from nested array and append gets incorrect value

### DISCUSSION THREAD

**Discussion by inbalserylevy** [react-hook-form/react-hook-form#4211] [Q&A]
https://user-images.githubusercontent.com/73095969/108097179-7b380a00-708a-11eb-9d46-c1f2772faeec.mov

1. I have 2 filter rows - (the 2 rows are set asynchronous from BE using reset):
2. I remove the first line from a field array (remove(index))
3. I append a new row (append({})
4. I get by default the previous values

I console.log my fields array and get the expected result:
![image](https://user-images.githubusercontent.com/73095969/108097532-ebdf2680-708a-11eb-9171-a32d79034814.png)

but inside my controllers, values are incorrect:
`      <Controller
        name={${namePrefix}[${index}].field}
        render={({ onChange, value, onBlur }) => {

          const fieldOptions = getRowOptions(filtersState.fieldsData, "field");
          return (
            <SelectInput
                name={`${namePrefix}[${index}].field`}
                onBlur={onBlur}
                size="m"
                onChange={(selectedValue) => {
                    setValue(`${namePrefix}[${index}].condition`, null);
                    setValue(`${namePrefix}[${index}].value`, null);
                    onChange(selectedValue.value);
                }}
                ref={register({required: "Required"})}
                placeholder={"+ Add Field"}
                options={fieldOptions}
                value={
                    fieldOptions?.filter(
                    (selectedField) => selectedField.value === value
                    )?.[0] || null
                    }
            />
          );
        }}
        rules={{ required: "Required" }}
        control={control}
      />`

**Comment by bluebill1049** [comment]
provide a codesandbox plz.

**Comment by inbalserylevy** [comment]
[https://codesandbox.io/s/react-hook-form-v6-controller-append-prepepend-insert-forked-zv5nu?file=/src/index.js](url)

remove the first "row" and "append" a new one.
It will get "cloud provider =" instead of an empty select inputs

**Reply by bluebill1049** [reply]
<img width="1462" alt="Screen Shot 2021-02-17 at 7 32 08 pm" src="https://user-images.githubusercontent.com/10513364/108177047-d5060600-7156-11eb-9b38-da40a0c9ee07.png">

**Reply by inbalserylevy** [reply]
https://codesandbox.io/s/react-hook-form-v6-controller-append-prepepend-insert-forked-obrx7


![image](https://user-images.githubusercontent.com/73095969/108177900-774ccc00-710c-11eb-952d-92c5c8bbdfb8.png)


Do I miss something?

**Reply by inbalserylevy** [reply]
condition.js content:


import React from "react";
import Row from "./row";
import { useFieldArray } from "react-hook-form";

const Condition = ({ namePrefix, control }) => {
  const { fields, append, remove } = useFieldArray({
    control, // control props comes from useForm (optional: if you are using FormContext)
    name: `${namePrefix}.conditions` // unique name for your Field Array
    // keyName: "id", default to "id", you can change the key name
  });

  const removeRow = (index) => {
    remove(index);
  };

  const addRow = () => {
    append({});
  };

  return fields.map((data, index, allData) => {
    return (
      <>
        <div key={data.id}>
          <Row
            index={index}
            onAddRow={addRow}
            onDeleteRow={removeRow}
            namePrefix={`${namePrefix}.conditions`}
            data={data}
          />
        </div>
      </>
    );
  });
};

export default Condition;

**Comment by inbalserylevy** [comment]
Another codesandbox which reproduces the bug: [https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-forked-4xf7o?fontsize=14&hidenavigation=1&theme=dark](url)


remove "cond1" and append to the same array.
The select gets previous value, while textbox is empty.
The problem is specific to "controller" items.

**Comment by inbalserylevy** [comment]
@bluebill1049 did you have the chance to take a look?

**Comment by bluebill1049** [comment]
<img width="755" alt="Screen Shot 2021-02-23 at 7 17 25 pm" src="https://user-images.githubusercontent.com/10513364/108817291-c498d400-760b-11eb-83a5-11f68143b55f.png">

why are you register `Controller`? it does the register for you.

**Reply by inbalserylevy** [reply]
I removed it already, but it doesn't seem to be related to the issue.

**Comment by inbalserylevy** [comment]
@bluebill1049 do you have any ideas? Is it an issue?

**Comment by hui21109** [comment]
encounter same issue here, any update?

**Reply by Moshyfawn** [reply]
Please, consider providing a reproduction codesanbox with the latest version of RHF

It's a bit of an old thread and quite a lot of package internals have changed. Asking if there has been any resolution doesn't really help the case

**Comment by iushev** [comment]
I have the same problem. 

Here is codesandbox:
https://codesandbox.io/s/old-hooks-ukfbwr?file=/src/App.tsx

When remove for example second email and immediately append new instead to added empty email it shows previously removed email.

When change appending to:
```javascript
...
          <button
            type="button"
            onClick={() => appnedEmail({ email: "", default: false })}
          >
            append
          </button>
...
``` 
it works fine.

For some reason if I didn't init email field with empty string it shows previously deleted

**Reply by Moshyfawn** [reply]
From [the RHF `useFieldArray` docs](https://react-hook-form.com/api/usefieldarray#shouldUnregister-example):
> **Important:** append data is required and not partial.

You need to provide the whole object to the `append` method. Partial data is not supported

**Reply by iushev** [reply]
I see. Thank you! :)

**Comment by ggmartins091** [comment]
I'm having the same problem as @iushev had, and I understand that I should pass the data when appending a new field.

My problem is that the appended value has a structure something like the following:
`{title: string, code: string, params: Record<string, string>}` where params has dynamic keys depending on which code is selected.

For example, if I select code 'A', I'll load a text input called 'url' which will be kept inside 'params' (params: {url: string}}, for code 'B' I have other fields inside params.

My problem is that when appending a new item, I don't know ahead of time which values 'params' will have, so I simply add an empty object, like this:
```
append({
      title: '',
      code: '',
      params: {},
});
```

The problem is the same as @iushev, if I delete the last item on the list, and then add a new one and select the same code, my params will have the same fieldvalues as the deleted one.

Any simple way to deal with this, other than knowing ahead of time what `code` is?

**Reply by Moshyfawn** [reply]
Would you consider creating a reproduction codesanbox?

**Reply by ggmartins091** [reply]
@Moshyfawn I tried to simplify stuff while still getting the same behaviour.

https://codesandbox.io/s/divine-star-00i37z?file=/src/App.tsx

To reproduce the problem I'm having, simply remove the second fieldArray, add another one  and select the code 'A' or 'C' (both have a property called url in common). You'll see that the url prior to removal will be shown, no matter the 'code', when there is a params.url property.

**Reply by Moshyfawn** [reply]
Thx, will take look!

**Reply by Moshyfawn** [reply]
I obviously don't know you particular use-case, but I'd suggest to provide a default code to append a new field with. After all, it's a single selection type UI, which in perfect world would always have one option selected _(kinda like radio buttons)_.

Additionally, you can introduce an empty state for your select and define a list of inputs _(or no inputs)_ to show on an empty state.

```js
  append({ title: "", code: "A", params: { url: "" } })
```

**Reply by ggmartins091** [reply]
Thanks for the reply!

You see, my problem is that "code" is something completely dynamic, meaning I can have someone create a new code with any params they want at any time, so I would need to always know every possible "param" for all codes and set all of them to an empty string even the ones that are not used for that selected code.

This is actually kinda of what I'm doing right now, just wanted to know if there was some easier way to append and ignore previous data.
